### ISSUE TITLE
Main thread blocks for a significant period when uploading a file

### ISSUE THREAD

**Issue by sdsantos** [blockstack/blockstack-android#151]
From what I was able to observe, the main thread blocks for a while before starting to upload a file on the network thread. The time it blocks is proportional to the size of the file being uploaded. On my Google Pixel 1 it gets to about 3-4 seconds for a 15MB file.

I couldn't find the cause yet, although I suspect it might be related with passing data around to V8. Any idea? Is this a known issue?

**Comment by friedger** [comment]
Are you calling `putFile` from within the v8thread?

**Comment by sdsantos** [comment]
Yes, which is the main thread. Otherwise it throws an error.

**Comment by friedger** [comment]
The default is that the v8 vm runs on the main thread. In the [example-service](https://github.com/blockstack/blockstack-android/blob/master/example-service/src/main/java/org/blockstack/android/BlockstackService.kt#L48) you can see how to move all this work to a dedicated thread.

**Comment by friedger** [comment]
This requires to implement your own executor.

**Comment by sdsantos** [comment]
Ok, interesting. I'll try that tomorrow to see if it works fine and solves
the issues.

**Comment by sdsantos** [comment]
@friedger It does work, and now I'm using that custom `Executor` all around the app to speed things up. 

The only extra complexity is that the BlockstackSession requires an Activity Context just for the login, and anywhere else the `ApplicationContext` is enough. It took a bit go around it, but it's all smooth now.

But I wonder, why isn't this the default implementation?

Thanks a lot for the help!

**Comment by friedger** [comment]
@sdsantos Good to hear that it worked.

It is exactly the login vs other actions problem. For login you can use the default implementation, for more complex applications it needs the custom solution. The more real use cases are emerging the better we can shape this SDK :-)

Do you have any suggestions how to make this easier to use? Something like a `LoginBlockstackSession`

**Comment by sdsantos** [comment]
I don't think another `BlockstackSession` is needed. Apparently we only need the Activity `Context` to call `startActivity`. Is that's correct, there's 2 simple solutions I can think of:

1. Call `context.startActivity` with the flag `Intent.FLAG_ACTIVITY_NEW_TASK` and it no longer needs to be an Activity `Context`, it can be the `ApplicationContext`.

2. Add a parameter to the `redirectUserToSignIn` method where the `Activity` is provided just for that `startActivity` call.

**Comment by friedger** [comment]
Indeed, only the `executor.onMainThread` needs an activity to start an activity.

Re 1. the user experience will suffer during login.
Re 2. one could add a parameter to set the current activity context when login. However, the idea is that the `BlockstackSession` belongs to an activity or service anyway.

Let's close this issue and discuss improvements of the activity handling in a new issue.

**Comment by sdsantos** [comment]
@friedger why would the user experience suffer with 1)?
And why should the `BlockstackSession` only belong to a single activity? Isn't it meant to be used throughout the app? I feel like belonging to the Application Context makes more sense.
